### PR TITLE
P4-2657 futher changes to include report date (a long with some refactoring) to audit imports.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
Changes:

- Fixing issues with the download spreadsheet where can use free text for supplier.
- Shifted the auditing of the spreadsheet download into the service so this is captured where it might take place.
- Shorted the build function names in the AuditableEvent.  
- Testing locally I spotted I had missed the report_date from the importing of reports. Also added missing test coverage.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/auditing/AuditableEvent.kt [hunk 1]
```diff
@@ -31,13 +31,13 @@ data class AuditableEvent(
       )
     }
 
-    fun createLogInEvent(authentication: Authentication) = createEvent(AuditEventType.LOG_IN, authentication)
+    fun logInEvent(authentication: Authentication) = createEvent(AuditEventType.LOG_IN, authentication)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/auditing/AuditableEvent.kt:34]: Don't think we really need create in the name.  When using the service we see something like this anyway:

`auditService.create(AuditableEvent.someEvent())`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/auditing/AuditableEvent.kt [hunk 2]
```diff
@@ -121,11 +121,12 @@ data class AuditableEvent(
       )
     }
 
-    fun createImportEvent(type: String, processed: Int, saved: Int) =
+    fun importReportEvent(type: String, reportDate: LocalDate, processed: Int, saved: Int) =
       createEvent(
         type = AuditEventType.REPORTING_DATA_IMPORT,
         metadata = mapOf(
           "type" to type,
+          "report_date" to reportDate.toString(),
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/auditing/AuditableEvent.kt:129]: I missed this one on the last PR.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/OutputSpreadsheetController.kt [hunk 3]
```diff
@@ -37,11 +37,9 @@ class OutputSpreadsheetController(
     response: HttpServletResponse?,
     authentication: Authentication?
   ): ResponseEntity<InputStreamResource?>? {
-    auditService.create(AuditableEvent.createDownloadSpreadsheetEvent(movesFrom, supplier, authentication!!))
-
-    return spreadsheetService.spreadsheet(supplier, movesFrom)?.let { file ->
+    return spreadsheetService.spreadsheet(authentication!!, Supplier.valueOfCaseInsensitive(supplier), movesFrom)?.let { file ->
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/OutputSpreadsheetController.kt:40]: This ensures we have a valid Supplier. Tthe value of UNKNOWN can be returned as well, probably OK for now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/OutputSpreadsheetController.kt [hunk 4]
```diff
@@ -37,11 +37,9 @@ class OutputSpreadsheetController(
     response: HttpServletResponse?,
     authentication: Authentication?
   ): ResponseEntity<InputStreamResource?>? {
-    auditService.create(AuditableEvent.createDownloadSpreadsheetEvent(movesFrom, supplier, authentication!!))
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/OutputSpreadsheetController.kt:40]: Shifted into service layer.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/SpreadsheetService.kt [hunk 5]
```diff
@@ -1,21 +1,27 @@
...
 
-  fun spreadsheet(supplierName: String, startDate: LocalDate): File? {
+  fun spreadsheet(authentication: Authentication, supplier: Supplier, startDate: LocalDate): File? {
+    logger.info("Generating spreadsheet for supplier '$supplier', moves from '$startDate''")
 
-    logger.info("Generating spreadsheet for supplier '$supplierName', moves from '$startDate''")
-
-    return pricesSpreadsheetGenerator.generate(Supplier.valueOfCaseInsensitive(supplierName), startDate)
+    return pricesSpreadsheetGenerator.generate(supplier, startDate).also {
+      auditService.create(AuditableEvent.downloadSpreadsheetEvent(startDate, supplier, authentication))
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/SpreadsheetService.kt:24]: Moved from controller to here.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/SpreadsheetServiceTest.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.verify
+import org.junit.jupiter.api.Test
+import org.springframework.security.core.Authentication
+import uk.gov.justice.digital.hmpps.pecs.jpc.auditing.AuditableEvent
+import uk.gov.justice.digital.hmpps.pecs.jpc.price.Supplier
+import uk.gov.justice.digital.hmpps.pecs.jpc.spreadsheet.PricesSpreadsheetGenerator
+import java.time.LocalDate
+
+internal class SpreadsheetServiceTest {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/SpreadsheetServiceTest.kt:12]: This service was lacking any test.

**Review by Thource — APPROVED** [review]

**Pull request merged**